File: app/src/main/java/com/xevenition/util/SaveUtil.java
package io.automile.automilepro.dagger.injectables;

import android.app.Activity;
import android.app.Application;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.automile.automilepro.interfaces.Preferenceable;

@Singleton
public class SaveUtil implements Preferenceable {
    private static final String APP_SHARED_PREFS = "io.automile.automilepro";
    public static final String ONBOARDING_MODEL_NOT_SELECTED = "ONBOARDING_MODEL_NOT_SELECTED";
    public static final String ONBOARDING_DESICION_NOT_DONE = "ONBOARDING_DESICION_NOT_DONE";
    public static final String TIP_TRIPS_SHOWN = "TIP_TRIPS_SHOWN";
    public static final String TIP_TASK_SHOWN = "TIP_TASK_SHOWN";
    public static final String TIP_VEHICLES_SHOWN = "TIP_VEHICLES_SHOWN";
    public static final String TIP_CHECKIN_SHOWN = "TIP_CHECKIN_SHOWN";
    public static final String KEY_UPSALE_SHOWN = "KEY_UPSALE_SHOWN";
    public static final String KEY_TRANSFER_INTERVAL = "KEY_TRANSFER_INTERVAL";
    public static final String KEY_SAMPLE_HARSH_EVENTS = "KEY_SAMPLE_HARSH_EVENTS";

    public static final String KEY_HOS_LAST_SYNC_DATE = "KEY_HOS_LAST_SYNC_DATE";
    public static final String KEY_HOS_UNSAVED_CHANGES = "KEY_HOS_UNSAVED_CHANGES";

    public static final String KEY_FILTER_EXPENSE = "KEY_FILTER_EXPENSE";
    public static final String KEY_FILTER_EXPENSE_ORDER = "KEY_FILTER_EXPENSE_ORDER";
    public static final String KEY_FILTER_EXPENSE_VEHICLE = "KEY_FILTER_EXPENSE_VEHICLE";
    public static final String KEY_FILTER_INSPECTION_ORDER = "KEY_FILTER_INSPECTION_ORDER";
    public static final String KEY_FILTER_INSPECTION_VEHICLE = "KEY_FILTER_INSPECTION_VEHICLE";
    public static final String KEY_FILTER_TASK = "KEY_FILTER_TASK";

    //Trip list filter settings
    public static final String KEY_FILTER_TRIP_DAYS = "KEY_FILTER_TRIP_DAYS";
    public static final String KEY_FILTER_TRIP_ORDER = "KEY_FILTER_TRIP_ORDER";
    public static final String KEY_FILTER_TRIP_VEHICLE = "KEY_FILTER_TRIP_VEHICLE";
    public static final String KEY_FILTER_TRIP_DRIVER = "KEY_FILTER_TRIP_DRIVER";
    public static final String KEY_FILTER_TRIP_DETAILS = "KEY_FILTER_TRIP_DETAILS";
    public static final String KEY_FILTER_TRIP_NOTES = "KEY_FILTER_TRIP_NOTES";

    public static final String KEY_FILTER_ORDER_GEO = "KEY_FILTER_ORDER_GEO";
    public static final String KEY_FILTER_ORDER_PLACES = "KEY_FILTER_ORDER_PLACES";

    public static final String KEY_VEHICLE_SORT_BY = "KEY_VEHICLE_SORT_BY";
    public static final String KEY_VEHICLE_ORDER_BY = "KEY_VEHICLE_ORDER_BY";


    public static final String USER_CONTACT_ID = "USER_CONTACT_ID";
    public static final String KEY_COLOR_1 = "KEY_COLOR_1";
    public static final String KEY_COLOR_2 = "KEY_COLOR_2";
    public static final String KEY_COLOR_3 = "KEY_COLOR_3";
    public static final String KEY_COLOR_4 = "KEY_COLOR_4";
    public static final String KEY_COLOR_5 = "KEY_COLOR_5";
    public static final String KEY_COLOR_6 = "KEY_COLOR_6";
    public static final String KEY_COLOR_7 = "KEY_COLOR_7";
    public static final String KEY_COLOR_8 = "KEY_COLOR_8";
    public static final String KEY_COLOR_9 = "KEY_COLOR_9";
    public static final String KEY_COLOR_10 = "KEY_COLOR_10";

    public static final String KEY_DEVICE_IDENTIFIER = "KEY_DEVICE_IDENTIFIER";
    public static final String KEY_MOVE_PUSH_SHOWN = "KEY_MOVE_PUSH_SHOWN";
    public static final String KEY_CUSTOM_CATEGORIES_SHOWN = "KEY_CUSTOM_CATEGORIES_SHOWN";

    //Live map filyter settings
    public static final String KEY_FILTER_LIVE_MOVING = "KEY_FILTER_LIVE_MOVING";
    public static final String KEY_FILTER_LIVE_PARKED = "KEY_FILTER_LIVE_PARKED";
    public static final String KEY_FILTER_LIVE_TRACKER = "KEY_FILTER_LIVE_TRACKER";
    public static final String KEY_FILTER_LIVE_VEHICLE = "KEY_FILTER_LIVE_VEHICLE";
    public static final String KEY_FILTER_LIVE_COLOR_1 = "KEY_FILTER_LIVE_COLOR_1";
    public static final String KEY_FILTER_LIVE_COLOR_2 = "KEY_FILTER_LIVE_COLOR_2";
    public static final String KEY_FILTER_LIVE_COLOR_3 = "KEY_FILTER_LIVE_COLOR_3";
    public static final String KEY_FILTER_LIVE_COLOR_4 = "KEY_FILTER_LIVE_COLOR_4";
    public static final String KEY_FILTER_LIVE_COLOR_5 = "KEY_FILTER_LIVE_COLOR_5";
    public static final String KEY_FILTER_LIVE_COLOR_6 = "KEY_FILTER_LIVE_COLOR_6";
    public static final String KEY_FILTER_LIVE_COLOR_7 = "KEY_FILTER_LIVE_COLOR_7";
    public static final String KEY_FILTER_LIVE_COLOR_8 = "KEY_FILTER_LIVE_COLOR_8";
    public static final String KEY_FILTER_LIVE_COLOR_9 = "KEY_FILTER_LIVE_COLOR_9";
    public static final String KEY_FILTER_LIVE_COLOR_10 = "KEY_FILTER_LIVE_COLOR_10";
    public static final String KEY_FILTER_JOB_DAYS = "KEY_FILTER_JOB_DAYS";
    public static final String KEY_FILTER_JOB_STATUS = "KEY_FILTER_JOB_STATUS";
    public static final String KEY_FILTER_JOBS_DRIVER = "KEY_FILTER_JOBS_DRIVER";
    public static final java.lang.String KEY_FILTER_JOB_PREVIOUS = "KEY_FILTER_JOB_PREVIOUS";
    public static final java.lang.String KEY_FILTER_JOB_UNASSIGNED = "KEY_FILTER_JOB_UNASSIGNED";
    public static final String KEY_FILTER_QUOTE_DAYS = "KEY_FILTER_QUOTE_DAYS";
    public static final String KEY_FILTER_QUOTE_STATUS = "KEY_FILTER_QUOTE_STATUS";
    public static final String KEY_FILTER_JOBREPORT_DAYS = "KEY_FILTER_JOBREPORT_DAYS";
    public static final String KEY_FILTER_JOBREPORT_DRIVER = "KEY_FILTER_JOBREPORT_DRIVER";
    public static final String KEY_FILTER_ANYTRACK_NOTIFICATIONS_DAYS = "KEY_FILTER_ANYTRACK_NOTIFICATIONS_DAYS";
    public static final String KEY_FILTER_ANYTRACK_NOTIFICATIONS_TYPE = "KEY_FILTER_ANYTRACK_NOTIFICATIONS_TYPE";

    private static SharedPreferences appSharedPrefs;

    private Editor prefsEditor;
    public static final String KEY_CHECKED_IN_VEHICLE = "KEY_CHECKED_IN_VEHICLE";
    public static final String KEY_ACCESS_TOKEN = "KEY_ACCESS_TOKEN";
    public static final String KEY_CLIENT_TOKEN = "KEY_CLIENT_TOKEN";
    public static final String KEY_REFRESH_TOKEN = "KEY_REFRESH_TOKEN";
    public static final String KEY_LOGIN = "KEY_LOGIN";
    public static final String KEY_USER_ID = "KEY_USER_ID";
    public static final String KEY_USER_CONTACT_ID = "KEY_USER_CONTACT_ID";
    public static final String KEY_AUTHENTICATION_EXPIRATION = "KEY_AUTHENTICATION_EXPIRATION";
    public final static String KEY_NUMBER_OF_STARTUPS = "KEY_NUMBER_OF_STARTUPS";

    @Inject
    public SaveUtil(Application app){
            appSharedPrefs = app.getSharedPreferences(APP_SHARED_PREFS, Activity.MODE_PRIVATE);
    }

    public void deleteAllBasedata(){
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.remove(KEY_FILTER_ORDER_GEO);
        prefsEditor.remove(KEY_FILTER_ORDER_PLACES);
        prefsEditor.remove(KEY_FILTER_TRIP_DAYS);
        prefsEditor.remove(KEY_FILTER_TRIP_NOTES);
        prefsEditor.remove(KEY_FILTER_TRIP_VEHICLE);
        prefsEditor.remove(KEY_FILTER_TRIP_ORDER);
        prefsEditor.remove(KEY_FILTER_TASK);
        prefsEditor.remove(KEY_CHECKED_IN_VEHICLE);
        prefsEditor.remove(KEY_FILTER_EXPENSE);
        prefsEditor.remove(KEY_FILTER_EXPENSE_ORDER);
        prefsEditor.remove(KEY_FILTER_EXPENSE_VEHICLE);
        prefsEditor.remove(KEY_FILTER_INSPECTION_ORDER);
        prefsEditor.remove(KEY_FILTER_INSPECTION_VEHICLE);
        prefsEditor.remove(KEY_ACCESS_TOKEN);
        prefsEditor.remove(KEY_REFRESH_TOKEN);
        prefsEditor.remove(KEY_CLIENT_TOKEN);
        prefsEditor.remove(KEY_LOGIN);
        prefsEditor.remove(KEY_USER_ID);
        prefsEditor.remove(KEY_AUTHENTICATION_EXPIRATION);
        prefsEditor.remove(ONBOARDING_MODEL_NOT_SELECTED);
        prefsEditor.remove(KEY_VEHICLE_ORDER_BY);
        prefsEditor.remove(KEY_VEHICLE_SORT_BY);
        prefsEditor.remove(KEY_MOVE_PUSH_SHOWN);
        prefsEditor.remove(KEY_FILTER_JOB_DAYS);
        prefsEditor.remove(KEY_FILTER_JOB_STATUS);
        prefsEditor.remove(KEY_FILTER_JOBS_DRIVER);
        prefsEditor.remove(KEY_FILTER_JOB_PREVIOUS);
        prefsEditor.remove(KEY_FILTER_JOB_UNASSIGNED);
        prefsEditor.remove(KEY_FILTER_QUOTE_STATUS);
        prefsEditor.remove(KEY_FILTER_QUOTE_DAYS);
        prefsEditor.remove(KEY_FILTER_JOBREPORT_DAYS);
        prefsEditor.remove(KEY_FILTER_JOBREPORT_DRIVER);
        prefsEditor.remove(KEY_FILTER_ANYTRACK_NOTIFICATIONS_DAYS);
        prefsEditor.remove(KEY_FILTER_ANYTRACK_NOTIFICATIONS_TYPE);
        prefsEditor.commit();
    }

    public void clearAllFiltersAtAppStart() {
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.remove(KEY_FILTER_EXPENSE);
        prefsEditor.remove(KEY_FILTER_EXPENSE_ORDER);
        prefsEditor.remove(KEY_FILTER_EXPENSE_VEHICLE);
        prefsEditor.remove(KEY_FILTER_INSPECTION_ORDER);
        prefsEditor.remove(KEY_FILTER_INSPECTION_VEHICLE);
        prefsEditor.remove(KEY_FILTER_TRIP_DAYS);
        prefsEditor.remove(KEY_FILTER_TRIP_NOTES);
        prefsEditor.remove(KEY_FILTER_TRIP_ORDER);
        prefsEditor.remove(KEY_FILTER_ORDER_GEO);
        prefsEditor.remove(KEY_FILTER_ORDER_PLACES);
        prefsEditor.remove(KEY_FILTER_TRIP_VEHICLE);
        prefsEditor.remove(KEY_FILTER_TASK);
        prefsEditor.remove(KEY_VEHICLE_SORT_BY);
        prefsEditor.remove(KEY_VEHICLE_ORDER_BY);
        prefsEditor.remove(KEY_COLOR_1);
        prefsEditor.remove(KEY_COLOR_2);
        prefsEditor.remove(KEY_COLOR_3);
        prefsEditor.remove(KEY_COLOR_4);
        prefsEditor.remove(KEY_COLOR_5);
        prefsEditor.remove(KEY_COLOR_6);
        prefsEditor.remove(KEY_COLOR_7);
        prefsEditor.remove(KEY_COLOR_8);
        prefsEditor.remove(KEY_COLOR_9);
        prefsEditor.remove(KEY_COLOR_10);
        prefsEditor.remove(KEY_FILTER_JOB_DAYS);
        prefsEditor.remove(KEY_FILTER_JOB_STATUS);
        prefsEditor.remove(KEY_FILTER_JOBS_DRIVER);
        prefsEditor.remove(KEY_FILTER_JOB_PREVIOUS);
        prefsEditor.remove(KEY_FILTER_JOB_UNASSIGNED);
        prefsEditor.remove(KEY_FILTER_QUOTE_STATUS);
        prefsEditor.remove(KEY_FILTER_QUOTE_DAYS);
        prefsEditor.remove(KEY_FILTER_JOBREPORT_DAYS);
        prefsEditor.remove(KEY_FILTER_JOBREPORT_DRIVER);
        prefsEditor.remove(KEY_FILTER_ANYTRACK_NOTIFICATIONS_DAYS);
        prefsEditor.remove(KEY_FILTER_ANYTRACK_NOTIFICATIONS_TYPE);
        prefsEditor.apply();
    }

    public void deleteDataForNewSignUp(){
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.remove(TIP_TRIPS_SHOWN);
        prefsEditor.remove(TIP_TASK_SHOWN);
        prefsEditor.remove(TIP_VEHICLES_SHOWN);
        prefsEditor.remove(TIP_CHECKIN_SHOWN);
        prefsEditor.remove(KEY_UPSALE_SHOWN);
        prefsEditor.commit();
    }

    @Override
    public String getString(String key) {
        return appSharedPrefs.getString(key, "");
    }

    @Override
    public long getLong(String key) {
        return 0;
    }

    @Override
    public void saveString(String key, String content) {
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.putString(key, content);
        prefsEditor.commit();
    }

    @Override
    public boolean getBoolean(String key) {
        return appSharedPrefs.getBoolean(key, false);
    }

    @Override
    public boolean getBoolean(String key, boolean defaultValue) {
        return appSharedPrefs.getBoolean(key, defaultValue);
    }

    @Override
    public void saveBoolean(String key, boolean content) {
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.putBoolean(key, content);
        prefsEditor.commit();
    }

    @Override
    public int getInt(String key, int defaultValue) {
        return appSharedPrefs.getInt(key, defaultValue);
    }

    @Override
    public double getDouble(String key) {
        return 0;
    }

    @Override
    public void saveInt(String key, int content) {
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.putInt(key, content);
        prefsEditor.commit();
    }

    @Override
    public void saveDouble(String key, double value) {

    }

    @Override
    public void saveLong(String key, long content){
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.putLong(key, content);
        prefsEditor.commit();
    }

    @Override
    public long getLong(String key, long defaultValue) {
        return appSharedPrefs.getLong(key, defaultValue);
    }

    @Override
    public void saveList(String key, ArrayList<String> list) {
        Set<String> set = new HashSet<String>();
        set.addAll(list);
        prefsEditor = appSharedPrefs.edit();
        prefsEditor.putStringSet(key, set);
        prefsEditor.commit();
    }

    @Override
    public ArrayList<String> getList(String key) {
        Set<String> set = appSharedPrefs.getStringSet(key, null);
        if (set == null)
            return null;
        return new ArrayList<String>(set);
    }
}

File: app/src/main/java/com/xevenition/util/SaveUtil.kt
package com.xevenition.util

import android.app.Activity
import android.app.Application
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import java.util.*
import javax.inject.Inject


class SaveUtil @Inject
constructor(app: Application) {

    companion object {
        val APP_SHARED_PREFS = "com.xevenition.util.homefinder"
        val KEY_GAME_ACTIVATED = "KEY_GAME_ACTIVATED"
        val KEY_GAME_NOT_PLAYING = "KEY_GAME_NOT_PLAYING"
        val KEY_MAP_MODE = "KEY_MAP_MODE"
        val KEY_MAP_MARKER_INFO = "KEY_MAP_MARKER_INFO"
        val KEY_FAVS = "KEY_FAVS"
        val KEY_INTRO_SEEN = "KEY_INTRO_SEEN"
        val KEY_SOLD_MODE = "KEY_SOLD_MODE"
        val KEY_FILTER_ON = "KEY_FILTER_ON"

        val PARAM_MIN_PRICE = "minPrice"
        val PARAM_MAX_PRICE = "maxPrice"
        val PARAM_MIN_SQUAREPRICE = "minSquarePrice"
        val PARAM_MAX_SQUAREPRICE = "maxSquarePrice"
        val PARAM_RENT = "maxRent"
        val PARAM_MIN_AREA = "minArea"
        val PARAM_MAX_AREA = "maxArea"
        val PARAM_MIN_PLOT = "minPlotArea"
        val PARAM_MAX_PLOT = "maxPlotArea"
        val PARAM_MIN_CONSTRUCTION = "minConstructionYear"

        val PARAM_MAX_CONSTRUCTION = "maxConstructionYear"
        val PARAM_DISTANCE_SEA = "maxDistanceToWater"
        val PARAM_ACTIVE_DAYS = "daysActive"

        val PARAM_SHOW_APARTMENT = "lägenhet"
        val PARAM_SHOW_VILLA = "villa"
        val PARAM_SHOW_TERRACED = "parhus"
        val PARAM_SHOW_HOLLIDAY = "fritidshus"
        val PARAM_SHOW_PLOT = "tomt/mark"
        val PARAM_SHOW_FARM = "gård"
        val PARAM_SHOW_ROW = "radhus"
        val PARAM_SHOW_SEMI = "kedjehus"

        val PARAM_SHOW_ROOM_1 = "1"
        val PARAM_SHOW_ROOM_2 = "2"
        val PARAM_SHOW_ROOM_3 = "3"
        val PARAM_SHOW_ROOM_4 = "4"
        val PARAM_SHOW_ROOM_5 = "5"
    }

    private var appSharedPrefs: SharedPreferences = app.getSharedPreferences(APP_SHARED_PREFS, Activity.MODE_PRIVATE)
    private var prefsEditor: Editor? = null

    fun getString(key: String): String {
        return appSharedPrefs.getString(key, "")
    }

    fun getLong(key: String): Long {
        return 0
    }

    fun saveString(key: String, content: String) {
        prefsEditor = appSharedPrefs.edit()
        prefsEditor!!.putString(key, content)
        prefsEditor!!.apply()
    }

    fun getBoolean(key: String): Boolean {
        return appSharedPrefs.getBoolean(key, false)
    }

    fun getBoolean(key: String, defaultValue: Boolean): Boolean {
        return appSharedPrefs.getBoolean(key, defaultValue)
    }

    fun saveBoolean(key: String, content: Boolean) {
        prefsEditor = appSharedPrefs.edit()
        prefsEditor!!.putBoolean(key, content)
        prefsEditor!!.apply()
    }

    fun getInt(key: String): Int {
        return appSharedPrefs.getInt(key, -1)
    }

    fun getInt(key: String, defaultValue: Int): Int {
        return appSharedPrefs.getInt(key, defaultValue)
    }

    fun getDouble(key: String): Double {
        return 0.0
    }

    fun saveInt(key: String, content: Int) {
        prefsEditor = appSharedPrefs.edit()
        prefsEditor!!.putInt(key, content)
        prefsEditor!!.apply()
    }

    fun saveDouble(key: String, value: Double) {

    }

    fun saveLong(key: String, content: Long) {
        prefsEditor = appSharedPrefs.edit()
        prefsEditor!!.putLong(key, content)
        prefsEditor!!.apply()
    }

    fun getLong(key: String, defaultValue: Long): Long {
        return appSharedPrefs.getLong(key, defaultValue)
    }

    fun saveList(key: String, list: ArrayList<String>) {
        val set = HashSet<String>()
        set.addAll(list)
        prefsEditor = appSharedPrefs.edit()
        prefsEditor!!.putStringSet(key, set)
        prefsEditor!!.apply()
    }

    fun getList(key: String): ArrayList<String>? {
        val set = appSharedPrefs.getStringSet(key, null) ?: return null
        return ArrayList(set)
    }

    fun isFilterActive(): Boolean {
        val minPrice = getInt(PARAM_MIN_PRICE)
        val maxPrice = getInt(PARAM_MAX_PRICE)
        val minArea = getInt(PARAM_MIN_AREA)
        val maxArea = getInt(PARAM_MAX_AREA)
        val minSqr = getInt(PARAM_MIN_SQUAREPRICE)
        val maxSqr = getInt(PARAM_MAX_SQUAREPRICE)
        val minPlot = getInt(PARAM_MIN_PLOT)
        val maxPlot = getInt(PARAM_MAX_PLOT)
        val minConstruct = getInt(PARAM_MIN_CONSTRUCTION)
        val maxConstruct = getInt(PARAM_MAX_CONSTRUCTION)
        val rent = getInt(PARAM_RENT)
        val distanceSea = getInt(PARAM_DISTANCE_SEA)
        val daysActive = getInt(PARAM_ACTIVE_DAYS)

        val showApartments = getBoolean(PARAM_SHOW_APARTMENT, true)
        val showVillas = getBoolean(PARAM_SHOW_VILLA, true)
        val showDetatched = getBoolean(PARAM_SHOW_TERRACED, true)
        val showCottage = getBoolean(PARAM_SHOW_HOLLIDAY, true)
        val showPlot = getBoolean(PARAM_SHOW_PLOT, true)
        val showFarms = getBoolean(PARAM_SHOW_FARM, true)
        val showSemi = getBoolean(PARAM_SHOW_SEMI, true)
        val showRow = getBoolean(PARAM_SHOW_ROW, true)

        val rooms1 = getBoolean(PARAM_SHOW_ROOM_1, true)
        val rooms2 = getBoolean(PARAM_SHOW_ROOM_2, true)
        val rooms3 = getBoolean(PARAM_SHOW_ROOM_3, true)
        val rooms4 = getBoolean(PARAM_SHOW_ROOM_4, true)
        val rooms5 = getBoolean(PARAM_SHOW_ROOM_5, true)
        if (showApartments && showVillas && showDetatched && showCottage && showPlot && showFarms && showSemi && showRow && rooms1 && rooms2 && rooms3 && rooms4 && rooms5) {
            if (minPrice == -1 && maxPrice == -1 && minArea == -1 && maxArea == -1 && minSqr == -1 && maxSqr == -1 && minPlot == -1 && maxPlot == -1 && minConstruct == -1 && maxConstruct == -1 && rent == -1 && distanceSea == -1 && daysActive == -1) {
                //filter is not active, do nothing
                return false
            } else {
                //filter is active
                return true
            }
        } else {
            //filter is active
            return true
        }
    }
}
